tls/s2n_ktls.c

```diff
@@ -15,11 +15,160 @@
 
 #include "tls/s2n_ktls.h"
 
+#include <sys/socket.h>
+
+#include "utils/s2n_socket.h"
+
 bool s2n_ktls_is_supported_on_platform()
 {
-#if defined(__linux__)
+#if S2N_KTLS_SUPPORTED
     return true;
 #else
     return false;
 #endif
 }
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    /* kTLS enable should only be called once the handshake has completed. */
+    if (!is_handshake_complete(conn)) {
+        RESULT_BAIL(S2N_ERR_HANDSHAKE_NOT_COMPLETE);
+    }
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
```

Comment:
Are we only enabling ktls for TLS1.2, or for any version <TLS1.3? Would we support ktls with TLS1.1?

Reply:
kTLS is only supported for 1.2 and 1.3 https://elixir.bootlin.com/linux/latest/source/net/tls/tls_main.c#L689